todoman/model.py

```diff
@@ -231,74 +276,472 @@ def _normalize_datetime(self, x):
+                FOREIGN KEY(file_id) REFERENCES files(id) ON DELETE CASCADE
+            );
+        ''')
+
+    def add_list(self, name, path, colour):
+        """
+        Inserts a new list into the cache.
+
+        Returns the id of the newly inserted list.
+        """
+
+        result = self.cur.execute(
+            'SELECT id FROM lists WHERE path = ?',
+            (path,),
+        ).fetchone()
+
+        if result:
+            return result['id']
+
+        try:
+            self.cur.execute(
+                "INSERT INTO lists (name, path, colour) VALUES (?, ?, ?)",
+                (name, path, colour,),
+            )
+        except sqlite3.IntegrityError as e:
+            raise Exception('Multiple lists named "{}"'.format(name)) from e
+
+        return self.add_list(name, path, colour)
+
+    def add_file(self, list_id, path, mtime):
+        self.cur.execute('''
+            INSERT OR IGNORE INTO files (
+                list_id,
+                path,
+                mtime
+            ) VALUES (?, ?, ?);
+            ''', (
+            list_id,
+            path,
+            mtime,
+        ))
+
+        result = self.cur.execute(
+            'SELECT id FROM files WHERE path = ?',
+            (path,),
+        ).fetchone()
+
+        return result['id']
+
+    def add_todo(self, todo, file_id):
+        """
+        Adds a todo into the cache.
+
+        :param icalendar.Todo todo: The icalendar component object on which
+        """
+
+        self.cur.execute('''
+            INSERT INTO todos (
+                file_id,
+                uid,
+                summary,
+                due,
+                priority,
+                created_at,
+                completed_at,
+                percent_complete,
+                dtstamp,
+                status,
+                description,
+                location,
+                categories
+            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
+            ''', (
+                file_id,
+                todo.uid,
+                todo.summary,
+                todo.due,
+                todo.priority,
+                todo.created_at,
+                todo.completed_at,
+                todo.percent_complete,
+                todo.dtstamp,
+                todo.status,
+                todo.description,
+                todo.location,
+                todo.raw_categories,
+            )
+        )
+
+    def todos(self, all=False, lists=[], urgent=False, location='',
+              category='', grep='', sort=[], reverse=False):
+        todos = []
+        list_map = self._lists_map()
+
+        extra_where = []
+        params = []
+
+        if not all:
+            extra_where.append('AND completed_at is NULL')
+        if lists:
+            q = ', '.join(['?'] * len(lists))
+
+            query = 'SELECT id FROM lists WHERE name IN ({});'.format(q)
+            list_ids = [row['id'] for row in self.conn.execute(q, lists)]
```

Comment:
Given this, wouldn't it make sense to make `list.name` the primary key? (not sure if possible in sqlite)